### PR TITLE
boostrapper: compact blocks before iterating them

### PR DESCRIPTION
**Pull request by a1k0n**
## Why this should be merged

Trying out @StephenButtolph's new interval tree based bootstrapping code, I found that it indeed fetched blocks much more quickly but that actually executing them was taking much longer than before. Ran a profile and found that most of the time was spent just iterating to the next block:

Screenshot from 2024-05-06 08-01-00

So as a quick experiment I hacked this in, and it dramatically reduced the ETA for execution. 

## How this works

Theory is that P chain blocks are stored in a pathological order on disk inside leveldb, despite the keys being contiguous, so spending a little time up front re-sorting the data blocks saves a lot of time later.

## How this was tested

Seat of my pants, really. But this is the profile afterwards:
image

I had left it running overnight, and the ETA this morning was about 6 hours. When I made this change and compacted my DB, the ETA dropped quickly down to ~50 minutes.

### CODE REVIEW

**Comment by a1k0n** [comment]
I guess this should have a check that there are more than ~5000 blocks to execute before bothering with compaction?

**Comment by StephenButtolph** [comment]
> I guess this should have a check that there are more than ~5000 blocks to execute before bothering with compaction?

I think it's fine... We're only compacting the block portion of the database... So if there are only a couple thousand blocks or so, I'd expect the compaction to be really fast.

File: snow/engine/snowman/bootstrap/interval/state.go
Comment (a1k0n) [reply snow/engine/snowman/bootstrap/interval/state.go]: Oh, right, since db is already a prefixdb. Sure, that makes sense.

**Comment by a1k0n** [comment]
> > I guess this should have a check that there are more than ~5000 blocks to execute before bothering with compaction?
> 
> I think it's fine... We're only compacting the block portion of the database... So if there are only a couple thousand blocks or so, I'd expect the compaction to be really fast.

Well, empirically it adds 2-3 seconds of startup time on every startup which I'd rather avoid, so I went ahead and added it.

File: snow/engine/snowman/bootstrap/interval/state.go
Comment (a1k0n) [reply snow/engine/snowman/bootstrap/interval/state.go]: Wait a second, does that actually work?
```
func (db *Database) Compact(start, limit []byte) error {
	db.lock.RLock()
	defer db.lock.RUnlock()

	if db.closed {
		return database.ErrClosed
	}

	prefixedStart := db.prefix(start)
	defer db.bufferPool.Put(prefixedStart)

	prefixedLimit := db.prefix(limit)
	defer db.bufferPool.Put(prefixedLimit)

	return db.db.Compact(*prefixedStart, *prefixedLimit)
}
```

If start and limit are both nil, isn't this an empty range?

Comment (a1k0n) [reply snow/engine/snowman/bootstrap/interval/state.go]: I think we would need something like this: ```
diff --git a/database/prefixdb/db.go b/database/prefixdb/db.go
index 0e203653a..001ab324d 100644
--- a/database/prefixdb/db.go
+++ b/database/prefixdb/db.go
@@ -189,7 +189,12 @@ func (db *Database) Compact(start, limit []byte) error {
        prefixedStart := db.prefix(start)
        defer db.bufferPool.Put(prefixedStart)

-       prefixedLimit := db.prefix(limit)
+       var prefixedLimit *[]byte
+       if limit == nil {
+               prefixedLimit = db.suffix()
+       } else {
+               prefixedLimit = db.prefix(limit)
+       }
        defer db.bufferPool.Put(prefixedLimit)

        return db.db.Compact(*prefixedStart, *prefixedLimit)
@@ -233,6 +238,21 @@ func (db *Database) prefix(key []byte) *[]byte {
        return prefixedKey
 }

+func (db *Database) suffix() *[]byte {
+       keyLen := len(db.dbPrefix)
+       prefixedKey := db.bufferPool.Get(keyLen)
+       copy(*prefixedKey, db.dbPrefix)
+       n := len(db.dbPrefix)
+       // lexically increment prefix to find the suffix of this db range
+       for i := n - 1; i >= 0; i-- {
+               (*prefixedKey)[i]++
+               if (*prefixedKey)[i] != 0 {
+                       break
+               }
+       }
+       return prefixedKey
+}
+
 // Batch of database operations
 type batch struct {
        database.Batch```

@StephenButtolph Am I wrong?

Comment (StephenButtolph) [reply snow/engine/snowman/bootstrap/interval/state.go]: I believe you are correct. Would you mind making a separate PR for this?

**Comment by a1k0n** [comment]
added the Compact fix to this PR; PTAL

**Comment by a1k0n** [comment]
Sorry, just saw your ask for a separate PR. Yeah, one sec.

File: snow/engine/snowman/bootstrap/interval/state.go
Comment (a1k0n) [reply snow/engine/snowman/bootstrap/interval/state.go]: Made. What a waste of PR #3000

File: snow/engine/snowman/bootstrap/storage.go
Comment (StephenButtolph) [inline snow/engine/snowman/bootstrap/storage.go:138]:
```suggestion
	totalNumberToProcess := tree.Len()
	if totalNumberToProcess > minBlocksToCompact {
```
(And then we can delete the `totalNumberToProcess` below here)

File: snow/engine/snowman/bootstrap/storage.go
Comment (StephenButtolph) [inline snow/engine/snowman/bootstrap/storage.go:139]: Super minor nit just to align the log a bit more with the following `executing blocks` log
```suggestion
		log("compacting database before executing blocks...")
```

**Review by StephenButtolph — COMMENTED** [review]
super minor nits then lgtm

**Comment by StephenButtolph** [comment]
Great catch! 🚀

**Review by StephenButtolph — COMMENTED** [review]
Going to run some perf tests on my end as well tonight - but code lgtm

File: snow/engine/snowman/bootstrap/storage.go
Comment (joshua-kim) [inline snow/engine/snowman/bootstrap/storage.go:144]:
```suggestion		
		if err := db.Compact(nil, nil); err != nil {
			// Not a fatal error, log and move on.
			log("failed to compact bootstrap database before executing blocks", zap.Error(err))
		}
```

**Review by joshua-kim — COMMENTED** [review]
Minor nit then LGTM

**Comment by a1k0n** [comment]
I'm doing some tests too, and I think maybe this PR is a bad idea, and just tuning the leveldb `writeBuffer` and `compactionTable*` options would work better. We have ~15GB of P chain blocks on mainnet, being continuously re-compacted into 2MB files. Even just downloading the initial P chain blocks is mostly blocked on all the re-compaction leveldb is doing, at least on my laptop here.

I also tried pebbledb but its performance was even worse.

Did you guys do much tuning on the defaults here?

**Comment by StephenButtolph** [comment]
> Did you guys do much tuning on the defaults here?

Not particularly. We do need to be careful with changing something like that to avoid any regressions in the (more important) normal operation of the node (rather than bootstrapping). Admittedly the verification logic is still happening though... So they might not be too different. Although bootstrapping is moving the blocks around whereas normally the blocks are written once by hash.

I'd be happy to merge this PR (once I can test the perf out on my node) and then we could look into optimizing level/pebble configs (we're hoping to move away from leveldb once we can get on a stable release of pebbledb).

**Review by joshua-kim — APPROVED** [review]

**Comment by StephenButtolph** [comment]
I tested this on my laptop and didn't see the same performance improvement. Going to spin up some AWS boxes to get a more reproducible benchmark.

**Comment by a1k0n** [comment]
Yeah, might have been particular to my setup which is why I'm not particularly enthusiastic about this change...

**Comment by StephenButtolph** [comment]
Did a run on an AWS c5.2xlarge box with 1 TiB of storage on a GP3 SSD (all defaults). I ended up executing the first 1.5m mainnet P-chain blocks.

The current tip of master took:
- `29m 11s` to execute
- ETA for completion was another `3h 51m 38s`
- Total ETA `4h 20m 49s`

This PR took:
- `4m 33s` to run the compaction
- `27m 28s` to execute
- ETA for completion was another `3h 37m 54s`.
- Total ETA `4h 9m 55s`

So, running the compaction did reduce the expected total time (after 1.5 M block executions) by `10m 54s` or `~4.2%`

@a1k0n what type of disk were you using?

**Comment by a1k0n** [comment]
> @a1k0n what type of disk were you using?

This was NVMe on a Linux laptop, ext4 filesystem. I have no idea why the performance was so variable. Happy to abandon this PR.

**Review by abi87 — APPROVED** [review]

**Comment by StephenButtolph** [comment]
We haven't seen a performance regression with this, and it seems like in some cases this can be a large performance improvement... So it makes sense to me to merge.

**Review by StephenButtolph — APPROVED** [review]

**Pull request merged**
